vortex-serde/src/layouts/read/layouts/flat.rs

```diff
@@ -0,0 +1,134 @@
+use std::cmp::min;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use vortex::compute::slice;
+use vortex::{Array, Context};
+use vortex_error::{vortex_err, VortexExpect, VortexResult, VortexUnwrap};
+use vortex_flatbuffers::footer;
+
+use crate::layouts::read::cache::RelativeLayoutCache;
+use crate::layouts::{
+    LayoutDeserializer, LayoutId, LayoutReader, LayoutSpec, Messages, ReadResult, Scan,
+};
+use crate::message_reader::ArrayBufferReader;
+use crate::stream_writer::ByteRange;
+
+#[derive(Debug)]
+pub struct FlatLayoutSpec;
+
+impl FlatLayoutSpec {
+    pub const ID: LayoutId = LayoutId(1);
+}
+
+impl LayoutSpec for FlatLayoutSpec {
+    fn id(&self) -> LayoutId {
+        Self::ID
+    }
+
+    fn layout(
+        &self,
+        fb_bytes: Bytes,
+        fb_loc: usize,
+        scan: Scan,
+        layout_serde: LayoutDeserializer,
+        message_cache: RelativeLayoutCache,
+    ) -> Box<dyn LayoutReader> {
+        let fb_layout = unsafe {
+            let tab = flatbuffers::Table::new(&fb_bytes, fb_loc);
+            footer::Layout::init_from_table(tab)
+        };
+        let flat_meta = fb_layout
+            .metadata()
+            .vortex_expect("FlatLayout must have metadata");
+        let begin = u64::from_le_bytes(
+            flat_meta.bytes()[0..8]
+                .try_into()
+                .map_err(|e| vortex_err!("Not a u64 {e}"))
+                .vortex_unwrap(),
+        );
+        let end = u64::from_le_bytes(
+            flat_meta.bytes()[8..16]
+                .try_into()
+                .map_err(|e| vortex_err!("Not a u64 {e}"))
+                .vortex_unwrap(),
+        );
+
+        Box::new(FlatLayout::new(
+            begin,
+            end,
+            scan,
+            layout_serde.ctx(),
+            message_cache,
+        ))
+    }
+}
+
+#[derive(Debug)]
+pub struct FlatLayout {
+    range: ByteRange,
+    scan: Scan,
+    ctx: Arc<Context>,
+    cache: RelativeLayoutCache,
+    done: bool,
+    cached_array: Option<Array>,
+}
+
+impl FlatLayout {
+    pub fn new(
+        begin: u64,
+        end: u64,
+        scan: Scan,
+        ctx: Arc<Context>,
+        cache: RelativeLayoutCache,
+    ) -> Self {
+        Self {
+            range: ByteRange { begin, end },
+            scan,
+            ctx,
+            cache,
+            done: false,
+            cached_array: None,
+        }
+    }
+
+    fn own_message(&self) -> Messages {
```

Comment:
WDYT about returning a single `Message` here? makes the length explicit

Reply:
it does, I can move it around.